### ISSUE TITLE
🐛[BUG] ModalForm 时在一定条件下会弹出失败 (和 SWR 相关) / ModalForm fails to pop up in certain scenarios (SWR-related)

### ISSUE THREAD

**Issue by tonywu6** [ant-design/pro-components#4184]
### 环境

`ant-design/pro-form@1.49.6`
`react@18.0.0-b`
`umi@3.5.20`
`react@18.0.0-b` and `react@16.14.0`
macOS 11.4
Chrome 96

### 描述

在使用 ProForm 的 ModalForm 时发现这样一个问题：当

1. ModalForm 使用 `request` 参数来异步加载表单数据
1. 页面上存在另一个 ProForm

当这另一个 ProForm 出于任何原因被卸载时，ModalForm 无法再次弹出，而是永远显示加载中。

### 复现代码

[**CodeSandbox**](https://codesandbox.io/s/proform-swr-global-caching-rerendering-fpdv9?file=/App.tsx)

### 暂时的解决方法

若将 ModalForm 包装在一个有[提供自己的缓存的](https://swr.vercel.app/docs/advanced/cache#create-cache-provider) `SWRConfig` 时问题会消失（见 CodeSandbox）

```tsx
import { ModalForm } from "@ant-design/pro-form";
import { SWRConfig } from "swr";

const MyForm = () => (
  <SWRConfig value={{ provider: () => new Map() }}>
    <ModalForm request={async () => ...}>
      ...
    </ModalForm>
  </SWRConfig>
)
```

---

#### 太长不看版

- ProForm 使用 [SWR](https://swr.vercel.app/) 进行表单数据缓存
- 页面中所有的 ProForm 表单共用同一个全局缓存（[SWR 默认行为](https://swr.vercel.app/docs/advanced/cache)）
- 出于某种原因，当任一个 ProForm 组件卸载时，[**它会清空全局缓存**](https://github.com/ant-design/pro-components/blob/6d5a40592da2b5342c508a99a9bfe0a42d87802a/packages/provider/src/index.tsx#L240-L250)（这应该不是期望的操作）
- 这导致了当 ModalForm 想要弹出、重新渲染、并想获取缓存了的表单数据时，SWR 返回 `undefined`
- ModalForm 认为 `undefined` 意味着数据还未加载，于是显示 spinner

#### 长版

默认情况下，SWR 对页面内所有的组件[提供一个全局的 cache](https://swr.vercel.app/docs/advanced/cache); 但是 SWR 也允许使用一个 `SWRConfig` 组件来对页面的一个区域进行参数自定义，其中包含了提供自己的 cache 的能力；也就是说，页面不同的区域可以使用单独的 cache:

```tsx
<SWRConfig value={{ provider: () => new Map() }}>
    <MyComponent />  {/* SWRConfig 里所有的子组件会使用上面这个 Map */}
</SWRConfig>
```

ProForm 存在这样的使用 `SWRConfig` 的操作；所有的表单都会使用一个 [`ConfigProviderWrap`](https://github.com/ant-design/pro-components/blob/6d5a40592da2b5342c508a99a9bfe0a42d87802a/packages/provider/src/index.tsx#L240-L250) 的组件来设置 SWR 选项：

https://github.com/ant-design/pro-components/blob/6d5a40592da2b5342c508a99a9bfe0a42d87802a/packages/provider/src/index.tsx#L249-L251

但是在 `ConfigProviderWrap` 里又存在这样一个逻辑：当 `ConfigProviderWrap` 所在的组件卸载时，清空一个 `cache` 对象：

https://github.com/ant-design/pro-components/blob/6d5a40592da2b5342c508a99a9bfe0a42d87802a/packages/provider/src/index.tsx#L240-L250

**问题在于，这里清空了的 `cache` 并不是 `ConfigProviderWrap` 里定义了的，即将被子组件使用的缓存, 而是 `ConfigProviderWrap` 的上级组件里定义的最接近的缓存 —— 若没有最接近的缓存，就会清空全局缓存。**

同时，这里声明了的缓存又并没有实际被表单使用，这是因为表单使用一个 `useFetchData` 来获取数据，这个 hook 会调用 `useSWR`，但是缓存声明存在于较 `useFetchData` 所在组件更内层的组件里：

```tsx
<ModalForm>
    <RequestForm>  {/* useFetchData 在这里 */}
        <BaseForm>
            <ConfigProviderWrap>  {/* cache 在这里 */}
```

于是乎，当页面中存在一个 ModalForm 和一个可以被卸载的 ProForm 时，

- 两个表单都在使用全局缓存
- 两个表单在卸载时都会清空全局缓存

当 ModalForm 想要重新渲染并去获取数据时，SWR 知道这个数据被缓存过并尝试去读取，但是因为缓存被清空了，它只能给到 `undefined`，而拿到 `undefined` 的 ProForm 认为数据还未加载成功，于是就会一直显示正在加载。

---

### Description

When using `ModalForm`, where

1. the modal form uses the `request` parameter to asynchronously fetch form data, and
1. there exists another `ProForm` in the page,

**when that `ProForm` component unmounts (for any reason), the `ModalForm` will no longer open when clicked and will show the spinner indefinitely.**

### To reproduce

[**CodeSandbox sample**](https://codesandbox.io/s/proform-swr-global-caching-rerendering-fpdv9?file=/App.tsx)

Essentially, have a `ModalForm` that uses `request` (even if it doesn't make network requests), then have another `ProForm` in the same page that can be conditionally mounted/unmounted, for example, with a switch.

The issue does not arise when using native antd `Form`s or `Modal`s, does not arise when the `ModalForm` doesn't use `request`, and does not arise when the other `ProForm` doesn't unmount.

### Hot-fix
This issue seems to go away if we wrap the `<ModalForm>` inside a `<SWRConfig>` component that [provides its own cache](https://swr.vercel.app/docs/advanced/cache#create-cache-provider):

```tsx
import { ModalForm } from "@ant-design/pro-form";
import { SWRConfig } from "swr";

const MyForm = () => (
  <SWRConfig value={{ provider: () => new Map() }}>
    <ModalForm request={async () => ...}>
      ...
    </ModalForm>
  </SWRConfig>
)
```

See more below.

#### tl;dr

- ProForm uses [SWR](https://swr.vercel.app/) for data caching (for loading form data with `request`);
- All ProForms on the same page shares a global SWR cache object, which is the [default behavior](https://swr.vercel.app/docs/advanced/cache)
- A side-effect defined in a `BaseForm` config provider component **attempts to clear the global whenever the form unmounts,** which seems to be unintended;
- One ProForm unmounting and clearing the cache causes SWR to return `undefined` when the ModalForm tries to get its cached data;
- ModalForm thinks `undefined` means data hasn't been loaded yet and displays the spinner, except the data will never arrive (because SWR has finished) and it will load forever.

#### Tracing the bug

We did some digging and we think it's caused by how ProForm is interacting with [SWR](https://swr.vercel.app/), which ProForm uses to cache form data.

Normally, SWR keeps data cached from all components on the same page [in a global `cache` object that SWR keeps for itself.](https://swr.vercel.app/docs/advanced/cache) However, it also allows you to provide SWR configurations regionally, by wrapping your components inside a `SWRConfig` component, where you can also declare a regional cache:

```tsx
<SWRConfig value={{ provider: () => new Map() }}>
    <MyComponent />  {/* If it needs caching, it will use the new Map above instead of the global one */}
</SWRConfig>
```

in which case everything **inside** `SWRConfig` will use the cache declared here, and everything outside of it will use the global cache.

ProForm seems to be using this functionality to provide each Form with its own `cache`:
in the `pro-provider` package, in `ConfigProviderWrap`

https://github.com/ant-design/pro-components/blob/6d5a40592da2b5342c508a99a9bfe0a42d87802a/packages/provider/src/index.tsx#L249-L251

In `ConfigProviderWrap`, before return, there is also a `useEffect` hook that says "clear the cache object before the component unmounts"

https://github.com/ant-design/pro-components/blob/6d5a40592da2b5342c508a99a9bfe0a42d87802a/packages/provider/src/index.tsx#L240-L250

I am not entirely sure what the purpose of clearing the cache is. However, here the `cache` object comes from the [`useSWRConfig`](https://swr.vercel.app/docs/global-configuration#access-to-global-configurations) hook, which will give us, among other things, the `cache` object currently in effect in whichever component is calling the hook.

This is **not** the `cache` object about to be used by the `SWRConfig` that will be rendered in this component, **but the `cache` from whichever nearest `SWRConfig` among the parent components — or if there isn't any, the global `cache`.** In other words, `ConfigProviderMap` tries to clear the parent/global `cache` every time it unmounts.

---

A second concern comes from the fact that the forms don't actually use the `cache` created in `ConfigProviderWrap` to keep requested data.

In `RequestForm`, data is fetched using ProForm's `useFetchData` hook, which calls `useSWR`:

https://github.com/ant-design/pro-components/blob/6d5a40592da2b5342c508a99a9bfe0a42d87802a/packages/form/src/BaseForm/index.tsx#L474-L478

But where `useFetchData` is called is outside of where the `cache` was created:

```jsx
<ModalForm>
    <RequestForm>  {/* useFetchData called here */}
        <BaseForm>
            <ConfigProviderWrap>  {/* cache declared here */}
```

Hence every ProForm on a page defaults to using the global cache, unless we manually put a `SWRConfig` around it.

---

In our issue, we have a ModalForm and another ProForm that unmounts. We now know both forms

- use the global `cache`
- try to clear the global `cache` when unmounted (because the `ConfigProviderWrap` is unmounted)

1. When one of the forms unmounts, it clears the cache.
1. Then, when the ModalForm is about to rerender, it tries to call the `request` method we gave it.
1. SWR kicks in and tries to read from the `cache` because it knew the data was cached.
1. **Except SWR finds `undefined` because the global cache was cleared.**
1. SWR gives `undefined` to ProForm, which then thinks data hasn't been loaded yet, and it will show a spinner:

https://github.com/ant-design/pro-components/blob/6d5a40592da2b5342c508a99a9bfe0a42d87802a/packages/form/src/BaseForm/index.tsx#L480-L486

---

By manually wrapping the `ModalForm` with a `SWRConfig` providing its own cache, the issue would go away because the form now uses its exclusive `cache` unaffected by other ProForms unmounting.

```tsx
  <SWRConfig value={{ provider: () => new Map() }}>
    <ModalForm request={async () => ...}>
      ...
    </ModalForm>
  </SWRConfig>
```

**Comment by chenshuai2144** [comment]
来个pr？
